ports/esp32/network_lan.c

```diff
@@ -116,63 +119,88 @@ STATIC mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
         }
     }
 
-    self->mdc_pin = machine_pin_get_id(args[ARG_mdc].u_obj);
-    self->mdio_pin = machine_pin_get_id(args[ARG_mdio].u_obj);
-    self->phy_power_pin = args[ARG_power].u_obj == mp_const_none ? -1 : machine_pin_get_id(args[ARG_power].u_obj);
-
     if (args[ARG_phy_addr].u_int < 0x00 || args[ARG_phy_addr].u_int > 0x1f) {
         mp_raise_ValueError(MP_ERROR_TEXT("invalid phy address"));
     }
 
     if (args[ARG_phy_type].u_int != PHY_LAN8720 &&
-        args[ARG_phy_type].u_int != PHY_TLK110 &&
-        args[ARG_phy_type].u_int != PHY_IP101) {
+        args[ARG_phy_type].u_int != PHY_IP101 &&
+        args[ARG_phy_type].u_int != PHY_RTL8201 &&
+        #if ESP_IDF_VERSION_MINOR >= 3      // KSZ8041 is new in ESP-IDF v4.3
+        args[ARG_phy_type].u_int != PHY_KSZ8041 &&
+        #endif
+        args[ARG_phy_type].u_int != PHY_DP83848) {
         mp_raise_ValueError(MP_ERROR_TEXT("invalid phy type"));
     }
 
-    if (args[ARG_clock_mode].u_int != -1 &&
-        args[ARG_clock_mode].u_int != ETH_CLOCK_GPIO0_IN &&
-        // Disabled due ESP-IDF (see modnetwork.c note)
-        // args[ARG_clock_mode].u_int != ETH_CLOCK_GPIO0_OUT &&
-        args[ARG_clock_mode].u_int != ETH_CLOCK_GPIO16_OUT &&
-        args[ARG_clock_mode].u_int != ETH_CLOCK_GPIO17_OUT) {
-        mp_raise_ValueError(MP_ERROR_TEXT("invalid clock mode"));
-    }
+    self->mdc_pin = machine_pin_get_id(args[ARG_mdc].u_obj);
+    self->mdio_pin = machine_pin_get_id(args[ARG_mdio].u_obj);
+    self->phy_power_pin = args[ARG_power].u_obj == mp_const_none ? -1 : machine_pin_get_id(args[ARG_power].u_obj);
+
+    eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
+    mac_config.smi_mdc_gpio_num = self->mdc_pin;
+    mac_config.smi_mdio_gpio_num = self->mdio_pin;
+    esp_eth_mac_t *mac = esp_eth_mac_new_esp32(&mac_config);
 
-    eth_config_t config;
+    eth_phy_config_t phy_config = ETH_PHY_DEFAULT_CONFIG();
+    phy_config.phy_addr = self->phy_addr;
+    phy_config.reset_gpio_num = self->phy_power_pin;
+    self->phy = NULL;
 
     switch (args[ARG_phy_type].u_int) {
-        case PHY_TLK110:
-            config = phy_tlk110_default_ethernet_config;
-            break;
         case PHY_LAN8720:
-            config = phy_lan8720_default_ethernet_config;
+            self->phy = esp_eth_phy_new_lan8720(&phy_config);
             break;
         case PHY_IP101:
-            config = phy_ip101_default_ethernet_config;
+            self->phy = esp_eth_phy_new_ip101(&phy_config);
+            break;
+        case PHY_RTL8201:
+            self->phy = esp_eth_phy_new_rtl8201(&phy_config);
             break;
+        case PHY_DP83848:
+            self->phy = esp_eth_phy_new_dp83848(&phy_config);
+            break;
+        case PHY_KSZ8041:
+            self->phy = esp_eth_phy_new_ksz8041(&phy_config);
```

Comment:
This will create a warning for IDF 4.1

Reply:
Ok, thank you. KSZ8041 is new in Version 4.3. I will change that.